Apps/Src/UpdateVelocity.c

```diff
@@ -1,67 +1,109 @@
-                car_state->DesiredMotorCurrent = 1.0f;
-            } else if(car_state->CRSet == REGEN) { // If in regen mode
-                switch(car_state->RegenButtonMode) {
-                    case REGEN_OFF:
-                        car_state->CRSet = ACCEL;
-                        car_state->RegenBrakeRate = 4;
-                        break;
-                    case RATE3:
-                        car_state->RegenBrakeRate = 3;
-                        break;
-                    case RATE2:
-                        car_state->RegenBrakeRate = 2;
-                        break;
-                    case RATE1:
-                        car_state->RegenBrakeRate = 1;
-                        break;
-                }
-                car_state->DesiredVelocity = (car_state->CurrentVelocity*car_state->RegenBrakeRate)/4;
-                car_state->DesiredMotorCurrent = 1.0f;
-            }
-        } else { // Not in regen or cruise control mode
-            car_state->CRSet = ACCEL;
-            car_state->CruiseControlEnable = OFF;
+    MotorController_Init();
+
+    static uint8_t prevBrakePedalPercent;
+
+    //Pressed refers to it being pressed before
+    static State cruzEnablePressed = OFF;
+    static State cruzSetPressed = OFF;    
+
+    //State refers to whether we consider the enable and set to be on or off
+    static State cruzEnableState = OFF;
+    static State cruzSetState = OFF;
+    
+    float desiredVelocity = 0;
+    float desiredMotorCurrent = 0;
+    while (1)
+    {
+        uint8_t accelPedalPercent = Pedals_Read(ACCELERATOR);
+        uint8_t brakePedalPercent = Pedals_Read(BRAKE);
+
+        // If the brake is changed, then the brake lights might need to be changed
+        if (brakePedalPercent != prevBrakePedalPercent)
+        {
+            OSSemPost(&LightsChange_Sem4, OS_OPT_POST_ALL, &err);
         }
 
-        if(car_state->CRSet == ACCEL) {
-            car_state->DesiredVelocity = MAX_VELOCITY;
-            car_state->DesiredMotorCurrent = convertPedaltoMotorPercent(car_state->AccelPedalPercent);
+        prevBrakePedalPercent = brakePedalPercent;
+
+        State regenPressed = Switches_Read(REGEN_SW);
+
+        //Active refers to it currently being pressed
+        State cruzEnableActive = Switches_Read(CRUZ_EN);
+        State cruzSetActive = Switches_Read(CRUZ_ST);
+
+        //The cruzEnableState is toggled on the rising edge of the button press
+        if(cruzEnableActive && !cruzEnablePressed){
+            cruzEnablePressed = ON;
+            cruzEnableState ^= cruzEnableState;
+        }else if(!cruzEnableActive){
+            cruzEnablePressed = OFF;
         }
 
-        // Delay of few milliseconds (10)
-        OSTimeDlyHMSM (0, 0, 0, 10, OS_OPT_TIME_HMSM_STRICT, &err);
+        //The cruzSetState is toggled on the rising edge of the button press
+        if(cruzSetActive && !cruzSetPressed){
+            cruzSetPressed = ON;
+            cruzSetState ^= cruzSetState;
```

Comment:
Same thing as above:
```suggestion
            cruzSetState = (State) ((int)cruzSetState ^ 1);
```